File: Minimum Falling path/Recursion.py
def minimumPath(arr,r,c,size):
    if r == size-1: return arr[r][c]

    res = float('inf')
    for last in range(size):
        if last !=  c:
            res = min(res,arr[r][c] + minimumPath(arr,r+1,last,size))
    
    return res

def minimumFallingSum(arr,size):
    ans = float('inf')
    for i in range(size):
        ans =  min(ans,minimumPath(arr,0,i,size))
    return ans 

if __name__ == '__main__':
    arr = [
        [1,2,3],
        [1,2,3],
        [1,2,3]
    ]
    size =  len(arr)

    print(minimumFallingSum(arr,size))

File: Minimum Falling path/Recursion.java
class Recursion {

    public static int MinimumFallingSum(int[][] arr,int r,int c,int size) {
        if(r == size-1) return arr[r][c];

        int result = Integer.MAX_VALUE;
        for(int i = 0; i < size; i ++) {
            if( i != c)
                result =  Math.min(result, arr[r][c] + MinimumFallingSum(arr, r+1, i, size));
        }

        return result;
    }
    public static void main(String[] args) {
        int[][] arr = {
            {1,2,3},
            {1,2,3},
            {1,2,3}
        };

        int size = arr.length;

        int maximum = Integer.MAX_VALUE;
        for(int i = 0 ; i < size; i ++) {
            maximum = Math.min(maximum, MinimumFallingSum(arr,0,i,size));
        }

        System.out.println(maximum);
    }
}